Source/Events.Store.MongoDB/Processing/Streams/StreamProcessorStateDiscriminatorConvention.cs

```diff
@@ -33,39 +33,56 @@ class StreamProcessorStateDiscriminatorConvention : IDiscriminatorConvention
 
         /// <summary>
         /// Gets the correct type when deserializing objects from  <see cref="AbstractStreamProcessorState"/> collection
-        /// depending on the "Partitioned" field, which is set by this class.
+        /// depending on the given nominal type or "Partitioned" field.
         /// </summary>
         /// <param name="bsonReader">A <see cref="IBsonReader"/>.</param>
         /// <param name="nominalType">The nominal type.</param>
         /// <returns>The actual wanted type.</returns>
         public Type GetActualType(IBsonReader bsonReader, Type nominalType)
         {
-            var bookmark = bsonReader.GetBookmark();
-            bsonReader.ReadStartDocument();
-            var actualType = nominalType;
-            while (bsonReader.ReadBsonType() != BsonType.EndOfDocument)
+            if (nominalType == typeof(StreamProcessorState))
             {
-                var fieldName = bsonReader.ReadName();
-                if (fieldName == "Partitioned")
+                return typeof(StreamProcessorState);
+            }
+            else if (nominalType == typeof(Partitioned.PartitionedStreamProcessorState))
+            {
+                return typeof(Partitioned.PartitionedStreamProcessorState);
+            }
+            else if (nominalType == typeof(AbstractStreamProcessorState))
```

Comment:
I don't really see how this is fixing the issue. How does Mongo decide what to pass along as the `nominalType`?

And also - if there is no `Paritioned` field on the BSON document, I would expect Mongo to throw an exception elsewhere later when it is trying to instantiate an `AbstractStreamProcessorState`.

Reply:
Added exceptions for a missing `Partitioned` field.